### PR TITLE
Refactor order structures

### PR DESCRIPTION
**Pull request by shubhamgoyal42**
## Why this should be merged
Refactors order structure in memory database.

### CODE REVIEW

File: plugin/evm/orderbook/memory_database.go [hunk 1]
```diff
@@ -396,51 +403,161 @@ func (db *InMemoryDatabase) GetAllOrders() []Order {
...
 	return allOrders
 }
 
+func (db *InMemoryDatabase) GetMarketOrders(market Market) []Order {
+	db.mu.RLock() // only read lock required
+	defer db.mu.RUnlock()
+
+	allOrders := []Order{}
+	longOrders := db.LongOrders[market]
+	if longOrders != nil {
```
Comment (atvanguard) [inline plugin/evm/orderbook/memory_database.go:418]: i think nil check is not required.

File: plugin/evm/orderbook/memory_database.go [hunk 2]
```diff
@@ -396,51 +403,161 @@ func (db *InMemoryDatabase) GetAllOrders() []Order {
...
 	}
 	return allOrders
 }
 
+func (db *InMemoryDatabase) GetMarketOrders(market Market) []Order {
+	db.mu.RLock() // only read lock required
+	defer db.mu.RUnlock()
+
+	allOrders := []Order{}
+	longOrders := db.LongOrders[market]
```
Comment (atvanguard) [inline plugin/evm/orderbook/memory_database.go:417]: can use `db.LongOrders[market]` directly

File: plugin/evm/orderbook/memory_database.go [hunk 3]
```diff
@@ -480,18 +597,26 @@ func (db *InMemoryDatabase) GetLongOrders(market Market, lowerbound *big.Int, bl
...
-				if _order := db.getCleanOrder(order, blockNumber); _order != nil {
-					longOrders = append(longOrders, *_order)
-				}
-			} else {
-				longOrders = append(longOrders, *order)
+
+	marketOrders := db.LongOrders[market]
+	if marketOrders == nil {
+		return longOrders
+	}
```
Comment (atvanguard) [inline plugin/evm/orderbook/memory_database.go:604]: nil check is redundant, please verify and if true remove from everywhere in this file. easier on the eyes

File: plugin/evm/orderbook/contract_events_processor.go [hunk 4]
```diff
@@ -263,24 +263,24 @@ func (cep *ContractEventsProcessor) handleIOCOrderBookEvent(event *types.Log) {
...
+				RawOrder:                &iocOrder,
+				Salt:                    iocOrder.Salt,
+				ReduceOnly:              iocOrder.ReduceOnly,
 				BlockNumber:             big.NewInt(int64(event.BlockNumber)),
 				OrderType:               IOC,
 			}
-			log.Info("IOCOrder/OrderPlaced", "order", limitOrder, "number", event.BlockNumber)
-			cep.database.Add(&limitOrder)
+			log.Info("IOCOrder/OrderPlaced", "order", order, "number", event.BlockNumber)
+			cep.database.Add(&order)
```
Comment (atvanguard) [inline plugin/evm/orderbook/contract_events_processor.go:283]: Please create a function `iocOrderToOrder`

**Review by atvanguard — COMMENTED** [review]
`sortLongOrders` and `sortShortOrders` are unused so code can be deleted.

File: plugin/evm/orderbook/memory_database.go [hunk 5]
```diff
@@ -396,51 +403,161 @@ func (db *InMemoryDatabase) GetAllOrders() []Order {
...
+	market := order.Market
 	order.LifecycleList = append(order.LifecycleList, Lifecycle{order.BlockNumber.Uint64(), Placed, ""})
-	db.OrderMap[order.Id] = order
+
+	var marketOrders []*Order
+	var position int
+	if order.PositionType == LONG {
+		marketOrders = db.LongOrders[market]
+		position = sort.Search(len(marketOrders), func(i int) bool {
+			priceDiff := order.Price.Cmp((marketOrders)[i].Price)
```
[the remark this answers is not in the document]
Comment (shubhamgoyal42) [reply plugin/evm/orderbook/memory_database.go]: removed now

File: plugin/evm/orderbook/memory_database.go (hunk 1 above)
Comment (shubhamgoyal42) [reply plugin/evm/orderbook/memory_database.go]: correct, it was not required. Removed

File: plugin/evm/orderbook/memory_database.go (hunk 3 above)
Comment (shubhamgoyal42) [reply plugin/evm/orderbook/memory_database.go]: removed

File: plugin/evm/orderbook/contract_events_processor.go (hunk 4 above)
Comment (shubhamgoyal42) [reply plugin/evm/orderbook/contract_events_processor.go:283]: keeping it like this cuz this is the only place we convert IOCOrder to Order

File: plugin/evm/orderbook/memory_database.go (hunk 2 above)
Comment (shubhamgoyal42) [reply plugin/evm/orderbook/memory_database.go]: yep, using directly now

File: plugin/evm/orderbook/memory_database.go [hunk 6]
```diff
@@ -396,51 +403,161 @@ func (db *InMemoryDatabase) GetAllOrders() []Order {
...
+		log.Error("In Delete - orderId does not exist in the database", "orderId", orderId.Hex())
+		metrics.GetOrRegisterCounter("delete_order_id_not_found", nil).Inc(1)
+		return
+	}
+
+	market := order.Market
+	if order.PositionType == LONG {
+		marketOrders := db.LongOrders[market]
+		for i, _order := range marketOrders {
+			if _order.Id == orderId {
```
[the remark this answers is not in the document]
Comment (shubhamgoyal42) [reply plugin/evm/orderbook/memory_database.go]: added for function to get the idx from a list of orders

File: plugin/evm/orderbook/memory_database.go [hunk 7]
```diff
@@ -1205,3 +1144,12 @@ func deepCopyTrader(order *Trader) *Trader {
...
 	}
 }
+
+func getOrderIdx(orders []*Order, orderId common.Hash) int {
+	for i, order := range orders {
+		if order.Id == orderId {
+			return i
+		}
+	}
+	return -1
```
Comment (atvanguard) [inline plugin/evm/orderbook/memory_database.go:1154]: will result in a panic

**Review by atvanguard — APPROVED** [review]

**Pull request merged**